accepted/file-enumeration.md

```diff
@@ -0,0 +1,238 @@
+            FindPredicate<TState> predicate,
+            // Only used if FindOptions.Recurse is set. Default is to always recurse.
+            FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None)
+    }
+
+    public static class Directory
+    {
+        public static IEnumerable<TResult> Enumerate<TResult, TState>(
+            string path,
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    public static class DirectoryInfo
+    {
+        public static IEnumerable<TResult> Enumerate<TResult, TState>(
+            FindTransform<TResult, TState> transform,
+            FindPredicate<TState> predicate,
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    /// <summary>
+    /// Used for processing and filtering find results.
+    /// </summary>
+    public ref struct RawFindData
+    {
+        // This will have private members that hold the native data and
+        // will lazily fill in data for properties where such data is not
+        // immediately available in the current platform's native results.
+
+		// The full path to the directory the current result is in
+        public string Directory { get; }
+
+        // The full path to the starting directory for enumeration
+        public string OriginalDirectory { get; }
+
+        // The path to the starting directory as passed to the enumerable constructor
+        public string OriginalUserDirectory { get; }
+
+        // Note: using a span allows us to reduce unneeded allocations
+        public ReadOnlySpan<char> FileName { get; }
+        public FileAttributes Attributes { get; }
+        public long Length { get; }
+
+        public DateTime CreationTimeUtc { get; }
+        public DateTime LastAccessTimeUtc { get; }
+        public DateTime LastWriteTimeUtc { get; }
+    }
+}
+```
+
+### Transforms & Predicates
+
+We'll provide common predicates transforms for building searches.
+
+``` C#
+namespace System.IO
+{
+    internal static partial class FindPredicates
+    {
+        internal static bool NotDotOrDotDot(ref RawFindData findData)
+        internal static bool IsDirectory(ref RawFindData findData)
+    }
+
+    public static partial class FindTransforms
+    {
+        public static DirectoryInfo AsDirectoryInfo(ref RawFindData findData)
+        public static FileInfo AsFileInfo(ref RawFindData findData)
+        public static FileSystemInfo AsFileSystemInfo(ref RawFindData findData)
+        public static string AsFileName(ref RawFindData findData)
+        public static string AsFullPath(ref RawFindData findData)
+    }
+}
+
+```
+
+### DosMatcher
```

Comment:
I assume you would have a RegexMatcher as well.

Reply:
Ideally. We might have to have less than ideal perf to start with if we don't get the Span overloads on Regex at first.

---
In reply to: [154807693](https://github.com/dotnet/designs/pull/24#discussion_r154807693) [](ancestors = 154807693)